fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBMetaDataStore.java

```diff
@@ -420,6 +425,89 @@ public void saveRecordMetaData(@Nonnull RecordMetaDataProto.MetaData metaDataPro
         context.asyncToSync(FDBStoreTimer.Waits.WAIT_SAVE_META_DATA, saveAndSetCurrent(metaDataProto));
     }
 
+    private CompletableFuture<RecordMetaDataProto.MetaData> loadCurrentProto() {
+        return loadCurrentSerialized().thenApply(serialized -> {
+            if (serialized == null) {
+                // TODO: Should we throw an error?
+                return null;
+            }
+            return parseMetaDataProto(serialized);
+        });
+    }
+
+    private void saveWithSplitHelper(RecordMetaData recordMetaData) {
+        byte[] serialized = recordMetaData.toProto().toByteArray();
+        SplitHelper.saveWithSplit(context, getSubspace(), CURRENT_KEY, serialized, null);
+        if (cache != null) {
+            cache.setCurrentVersion(context, recordMetaData.getVersion());
+            addPendingCacheUpdate(recordMetaData);
+            addPendingCacheUpdate(serialized);
+        }
+    }
+
+    /**
+     * Add new index to the record meta-data.
+     * @param recordType the name of the record type
+     * @param indexName the name of the new index
+     * @param fieldName the field to be indexed
+     */
+    public void addIndex(@Nonnull String recordType, @Nonnull String indexName, @Nonnull String fieldName) {
+        addIndex(recordType, new Index(indexName, fieldName));
+    }
+
+    /**
+     * Add new index to the record meta-data.
+     * @param recordType the name of the record type
+     * @param indexName the name of the new index
+     * @param indexExpression the root expression of the index
+     */
+    public void addIndex(@Nonnull String recordType, @Nonnull String indexName, @Nonnull KeyExpression indexExpression) {
+        addIndex(recordType, new Index(indexName, indexExpression));
+    }
+
+    private void addIndex(@Nonnull String recordType, @Nonnull Index index) {
+        CompletableFuture<Void> future = loadCurrentProto().thenApply(metaDataProto -> {
+            RecordMetaDataBuilder recordMetaDataBuilder = createMetaDataBuilder(metaDataProto);
+            recordMetaDataBuilder.addIndex(recordType, index);
+            recordMetaData = recordMetaDataBuilder.build(true);
+            saveWithSplitHelper(recordMetaData);
```

Comment:
I can do it. I didn't call `saveAndSetCurrent` or `saveRecordMetaData` because it unnecessarily loads. If it's cached, probably it's fine. but when it's not there yet, we would unnecessarily do a look up. Maybe I can keep this, and then later change `saveAndSetCurrent` or `saveRecordMetaData` to call this too? What do you think?

Fixed the `maintainHistory` issue.

Reply:
Actually thinking thru this again. In order to properly keep history, probably it would better to call saveAndSetCurrent. I'll apply your comment. thanks.